tidy3d/components/medium.py

```diff
@@ -2927,6 +2928,256 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
     )
 
 
+""" Medium perturbation classes """
+
+
+class AbstractPerturbationMedium(ABC, Tidy3dBaseModel):
+    """Abstract class for medium perturbation."""
+
+    @abstractmethod
+    def perturbed_copy(
+        self,
+        temperature: SpatialDataArray = None,
+        electron_density: SpatialDataArray = None,
+        hole_density: SpatialDataArray = None,
+    ) -> Union[AbstractMedium, AbstractCustomMedium]:
+        """Sample perturbations on provided heat and/or charge data and create a custom medium.
+        Any of ``temperature``, ``electron_density``, and ``hole_density`` can be ``None``.
+        If all passed arguments are ``None`` then a non-custom medium is returned.
+        All provided fields must have identical coords.
+
+        Parameters
+        ----------
+        temperature : SpatialDataArray = None
+            Temperature field data.
+        electron_density : SpatialDataArray = None
+            Electron density field data.
+        hole_density : SpatialDataArray = None
+            Hole density field data.
+
+        Returns
+        -------
+        Union[AbstractMedium, AbstractCustomMedium]
+            Medium specification after application of heat and/or charge data.
+        """
+
+
+class PerturbationMedium(Medium, AbstractPerturbationMedium):
+    """Dispersionless medium with perturbations.
+
+    Example
+    -------
+    >>> from tidy3d import ParameterPerturbation, LinearHeatPerturbation
+    >>> dielectric = PerturbationMedium(
+    >>>     permittivity=4.0,
+    >>>     permittivity_perturbation=ParameterPerturbation(
+    >>>         heat=LinearHeatPerturbation(temperature_ref=300, coeff=0.0001),
+    >>>     ),
+    >>>     name='my_medium'
+    >>> )
+    """
+
+    permittivity_perturbation: Optional[ParameterPerturbation] = pd.Field(
+        ...,
+        title="Permittivity Perturbation",
+        description="List of heat and/or charge perturbations to permittivity.",
+        units=PERMITTIVITY,
+    )
+
+    conductivity_perturbation: Optional[ParameterPerturbation] = pd.Field(
+        None,
+        title="Permittivity Perturbation",
+        description="List of heat and/or charge perturbations to permittivity.",
+        units=CONDUCTIVITY,
+    )
+
+    _permittivity_perturbation_validator = validate_parameter_perturbation(
+        "permittivity_perturbation",
+        "permittivity",
+        allowed_real_range=[(1.0, None)],
+        allowed_imag_range=[None],
+        allowed_complex=False,
+    )
+    _conductivity_perturbation_validator = validate_parameter_perturbation(
```

Comment:
so, these validators only warn if conductivity/permittivity can potentially go out of bounds (within the provided/calculated `temperature_range`, `electron_range`, and `hole_range`). The strict validation is left to the corresponding `CustomMedium`. So perhaps it's still ok to warn the user about potentially negative conductivity even if `allow_gain=True`?

Reply:
I see, so it will warn and then error. In that regard, there seems to be missing `allow_gain` in the return of 
```
    return CustomMedium(
            permittivity=permittivity_field,
            conductivity=conductivity_field,
            name=self.name,
            subpixel=True,
        )
```
and it seems better to add `subpixel` as a field of PerturbedMedium so that one can choose to apply subpixel or not. But warn that for a uniform medium, supixel parameter will be ignored.

Reply:
Oh, you're right, thanks for catching that. Actually, `frequency_range` should also be passed. Fixed it.